fe/fe-core/src/main/java/org/apache/doris/qe/PointQueryExecutor.java

```diff
@@ -159,23 +191,181 @@ public static void directExecuteShortCircuitQuery(StmtExecutor executor,
+            return expr instanceof SlotRef && ((SlotRef) expr).getColumnName().equalsIgnoreCase(Column.DELETE_SIGN);
+        }
+    }
+
+    private static void updateScanNodeConjuncts(OlapScanNode scanNode, List<Expr> conjunctVals) {
+        List<Expr> conjuncts = scanNode.getConjuncts();
+        Map<Class<? extends Expr>, ConjunctHandler> handlers = Maps.newHashMap();
+        handlers.put(InPredicate.class, new InPredicateHandler());
+        handlers.put(BinaryPredicate.class, new BinaryPredicateHandler());
+        int conjunctValsBegin = 0;
+        for (Expr expr : conjuncts) {
+            ConjunctHandler handler = handlers.get(expr.getClass());
+            if (handler == null) {
+                throw new AnalysisException("Not support conjunct type " + expr.getClass().getName());
+            }
+            conjunctValsBegin = handler.handle(expr, conjunctVals, conjunctValsBegin);
+        }
+
+        Preconditions.checkState(conjunctValsBegin == conjunctVals.size());
+    }
+
+    public static byte[] appendBytes(byte[] original, byte[] toAppend) {
+        byte[] result = new byte[original.length + toAppend.length];
+        System.arraycopy(original, 0, result, 0, original.length);
+        System.arraycopy(toAppend, 0, result, original.length, toAppend.length);
+        return result;
     }
 
     public void setTimeout(long timeoutMs) {
         this.timeoutMs = timeoutMs;
     }
 
+    void getAllKeyTupleCombination(List<Expr> conjuncts, int index,
+                        List<String> currentKeyTuple,   // current key tuple
+                        List<List<String>> result,
+                        List<String> columnExpr,
+                        List<Column> keyColumns) {
+        if (index == conjuncts.size()) {
+            List<String> orderedKeyTuple = new ArrayList<>(currentKeyTuple.size());
+            OlapTable olapTable = shortCircuitQueryContext.scanNode.getOlapTable();
+
+            // add key tuple in keys order
+            for (Column column : keyColumns) {
+                int colIdx = columnExpr.indexOf(column.getName());
+                String currentKey = currentKeyTuple.get(colIdx);
+                orderedKeyTuple.add(currentKey);
+            }
+            List<String> keyTupleForDistributionPrune =
+                    Lists.newArrayList(distributionKeyColumns.stream().map((idx) -> {
+                        return orderedKeyTuple.get(idx);
+                    }).collect(Collectors.toList()));
+            result.add(Lists.newArrayList(orderedKeyTuple));
+            Set<Long> tabletIDs = Sets.newHashSet();
+            for (PartitionKey key : distributionKeys2TabletID.keySet()) {
+                List<String> distributionKeys = Lists.newArrayList();
+                for (LiteralExpr expr : key.getKeys()) {
+                    distributionKeys.add(expr.getStringValue());
+                }
+                if (distributionKeys.equals(keyTupleForDistributionPrune)) {
```

Comment:
if not equal, tabletIDs will be empty, and distributionKeys2TabletID is use less?

Reply:
There is always a keyTupleForDistributionPrune that is equal to the distributionKeys.